---
title: "Distinguish endothelial cells and pericytes in BA4 endo-pericyte populations"
author: "B180131"
date: "21/05/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is the scRNA-seq analysis workflow for comparing human brain endothelial cells vs human spinal cord endothelial cells. This allows for better understanding of how neurovascular functions are contributed by selected populations of EC populations. Required packages involve Seurat and Bioconductor. 

* This script attempts to characterize endothelial cells and pericytes within populations of both
BA4.

```{r, include=FALSE}
load_lib <- "Load Libraries"
```
## `r load_lib`
* Load all the necessary libraries


```{r load_lib, message=FALSE}
library(Seurat) 
library(dplyr)
library(scater)
library(SingleCellExperiment)
library(here)
library(ggplot2)
library(scran)
library(stringr)
library(clustree)
```


```{r, include=FALSE}
load_data <- "Load BA4 & CSC endothelial-pericyte populations"
```
## `r load_data`
* Load the QC-ed and normalised EC-dataset of BA4 populations from Luise's dataset.
* Here we save the BA4 endo-pericyte_1 and endo-pericyte_2 populations into a SeuratObjects for clustering.


```{r load_data, echo=TRUE}
BA4_CSC <- readRDS(here("data/BA4_CSC_QC.RDS")) #Load entire dataset
Idents(BA4_CSC) <- "clusters_named" #Set identity to clusters_named column on metadata
BA4 <- subset(x = BA4_CSC, subset=Tissue=="BA4")
```


```{r, include=FALSE}
cluster_EC_Peri <- "Distinguish Endothelial Cells and Pericytes in BA4 Endo-Pericyte Population"
```
## `r cluster_EC_Peri`
Aims to distinguish between endothelial cells and pericytes of BEC (BA4), through feature selection, dimension reduction, and clustering analysis.

* First identified the top 2000 highly variable genes in the endo-pericyte_1 & 2 populations through a local polynomial regression (loess)
* Scale data using a linear model
* Perform PCA
* Use the first 20 principal components for clustering analysis
* FindNeighbors (identify cell to cell connections using KNN)
* FindClusters (deconvolute graph using Louvain algorithm)
* Generate UMAP (at resolutions of 0.05, 0.1, and 0.2)


```{r cluster_EC_Peri, message=FALSE}
#Calculate based on variance of log-expression of genes across cells. 2000 HVGs are selected.
BA4 <- FindVariableFeatures(BA4, selection.method = "vst", nfeatures = 2000)
top10_vg <- head(VariableFeatures(BA4), 10)
top10_vg
```


###################### Using Bioconductor to calculate top HVGs #############################
```{r}
BA4_sce <- as.SingleCellExperiment(BA4)
dec_BA4 <- modelGeneVar(BA4_sce)

#Visualizing the fit:
fit.BA4 <- metadata(dec_BA4)
plot(fit.BA4$mean, fit.BA4$var, xlab="Mean of log-expression", ylab= "Variance of log-expression")
curve(fit.BA4$trend(x), col="dodgerblue", add=TRUE, lwd=2)
```

Identify top HVGs using Bioconductor
```{r}
HVG_sce <- getTopHVGs(dec_BA4, n=2000)
head(HVG_sce, 10)
```

###########################################################################################


Plot HVGs from Seurat with and without labels
```{r, fig.width = 15, fig.height= 8, message= FALSE, warning=FALSE}
var_genes <- VariableFeaturePlot(BA4)
labels <- LabelPoints(plot = var_genes, points = top10_vg, repel = TRUE)
var_genes + labels
```



```{r, include=FALSE}
PCA <- "Perform PCA"
```
## `r PCA`
Scale the data 

* Shifts the expression of each gene, so that the mean expression across cells is 0
* Scales the expression of each gene, so that the variance across cells is 1


        
```{r, message=FALSE}
all.genes <- rownames(BA4)
BA4 <- ScaleData(BA4, features = all.genes) #save all variable genes 
```



Perform linear dimension reduction (PCA). Assumes that biological process affects multiple genes in a coordinated manner. PCs are likely to represent biological structure as more variation can be captured by considering the correlated behavior of many genes. 
```{r, message=FALSE}
#Perform PCA on log-normalised expression values of genes. Number of genes for PCA is based on the HVGs selected above.
BA4 <- RunPCA(BA4, features = VariableFeatures(object=BA4))
```



```{r, include=FALSE}
visualize <- "Visualization"
```
## `r visualize`
View the top genes associated with principal components 1 and 2.


```{r, fig.width= 10, fig.height= 8}
VizDimLoadings(BA4, dims = 1:2, reduction = "pca") 
```


Graph output of a PCA dimension reduction on 2D scatter plot. X and Y axes represent the top two PCs. Each point is a cell, and it's position decided based on the variance of expression of genes within the PCs. Cells are colored either as endo-pericyte_1 or endo-pericyte_2.
```{r}
DimPlot(BA4, reduction = "pca")
```


Draws a heatmap focusing on PC1. Both cells and genes are sorted by their principal component scores. Shows gene heterogeneity of PC1 in the top 500 cells.
```{r}
DimHeatmap(BA4, dims = 1, cells= 500, balanced = TRUE)
```


Visualize the heatmap of first 20 principal components in the top 500 cells.
```{r, fig.width=10, fig.height=25, fig.fullwidth= TRUE}
DimHeatmap(BA4, dims = 1:20, cells= 500, balanced = TRUE)
```


Plots the standard deviations of the PCs for easy identification of their significance.

* 20 PCs will be used for downstream analyses.
```{r}
ElbowPlot(BA4, ndims = 50)
```


```{r, include=FALSE}
cluster <- "Clustering Analysis"
```
Cluster of Endo-pericyte_1 & 2 BA4 populations

* Cluster to distinguish endothelial cell and pericyte populations
* Cluster resolution attempted at 0.05, 0.1, and 0.2

t-SNE and UMAP gives more freedom to how it arranges cells in low dimensional space. Allowing the separation of many clusters in a complex population than PCA. Each cell is connected to it's nearest neighboring cells (KNN). Louvain algorithm then deconvolutes connections to identify clusters.


```{r, message=FALSE, warning= FALSE}
BA4 <- FindNeighbors(BA4, dims = 1:20) #KNN
BA4 <- FindClusters(BA4, resolution = 0.05) #Louvain algorithm
Idents(BA4) <- "RNA_snn_res.0.05"
```


```{r, message=FALSE, warning= FALSE}
##Plot UMAP and t-SNE
BA4 <- RunUMAP(BA4, dims = 1:20)
BA4 <- RunTSNE(BA4, dims = 1:20)

DimPlot(BA4, reduction = "umap", label = T)
DimPlot(BA4, reduction = "tsne", label = T)
DimPlot(BA4, reduction = "pca", label = F, pt.size = 2)
```


```{r, message=FALSE, warning= FALSE}
BA4 <- FindClusters(BA4, resolution = 0.1) #Louvain algorithm
Idents(BA4) <- "RNA_snn_res.0.1"
```


```{r, message=FALSE, warning = FALSE}
##Plot UMAP and t-SNE
BA4 <- RunUMAP(BA4, dims = 1:20)
BA4 <- RunTSNE(BA4, dims = 1:20)

DimPlot(BA4, reduction = "umap", label = T) #Generates more compact clusters with more empty spaces between them. Faster than t-SNE
DimPlot(BA4, reduction = "tsne", label = T) #Computationally intensive than other visualizations
DimPlot(BA4, reduction = "pca", label = F, pt.size = 2)
```


```{r, message=FALSE, warning = FALSE}
BA4 <- FindClusters(BA4, resolution = 0.5) #Louvain algorithm
Idents(BA4) <- "RNA_snn_res.0.5"
```


```{r, message=FALSE, warning = FALSE}
##Plot UMAP and t-SNE
BA4 <- RunUMAP(BA4, dims = 1:20)
BA4 <- RunTSNE(BA4, dims = 1:20)

DimPlot(BA4, reduction = "umap", label = T)
DimPlot(BA4, reduction = "tsne", label = T)
DimPlot(BA4, reduction = "pca", label = F, pt.size = 2)
```


Set identity to resolution 0.1
```{r}
Idents(BA4) <- "RNA_snn_res.0.05"
```


```{r, include=FALSE}
cluster_annotation <- "Cluster annotation"
```
## `r cluster_annotation`
How many cells contribute to each cluster.


```{r, echo = TRUE}
CellsPerCluster <- as.data.frame(tapply(
    BA4@meta.data$Barcode, BA4@meta.data$RNA_snn_res.0.05, function(x) length(x) ))

names(CellsPerCluster) <- "NumberOfCells"
CellsPerCluster$RNA_snn_res.0.1 <- rownames(CellsPerCluster)
CellsPerCluster$Cluster <- rownames(CellsPerCluster)
CellsPerCluster
```


Find markers that define clusters via differential expression. Compares cluster 0 to cluster 1. T-test returns a list of candidate markers for cluster 0.
```{r}
clusters.marker <- FindMarkers(BA4, ident.1 = 0, ident.2 = 1, min.pct = 0.25)
```


Upregulated genes in cluster 0
```{r}
upregulated_EC <- subset(clusters.marker, clusters.marker$avg_log2FC > 0)
head(upregulated_EC, 50)
```


Upregulated genes in cluster 1
```{r}
clusters1.marker <- FindMarkers(BA4, ident.1 = 1, ident.2 = 0, min.pct = 0.25)
upregulated_Peri <- subset(clusters1.marker, clusters1.marker$avg_log2FC > 0) #Reverse of log-FC
head(upregulated_Peri, 50)
```

The following markers were used to distinguish cell types:

* endothelial cells: VWF, FLT1, CLDN5, ABCB1, THSD4, PECAM1, CDH5
* pericytes: PDGFRB, NOTCH3, PRKG1, CSPG4 (NG2)

```{r, include=FALSE}
feature_plt_1 <- "Feature Plots (EC genes)"
```
## `r feature_plt_1`

```{r feature_plt_1, fig.width=15, fig.height=13}

#EC  Biomarkers
FeaturePlot(BA4, features = c("VWF",
                              "FLT1",
                              "CLDN5",
                              "ABCB1",
                              "ST6GALNAC3",
                              "ELOVL7",
                              "THSD4",
                              "PECAM1",
                              "MECOM"
                              ))
```


```{r, fig.width= 7, fig.height=10}
VlnPlot(BA4, features = c("VWF", "FLT1", "CLDN5", "ABCB1",  "ST6GALNAC3", "ELOVL7", "THSD4", "PECAM1", "MECOM"))
```


```{r, include=FALSE}
feature_plt_2 <- "Feature Plots (Pericyte genes)"
```
## `r feature_plt_2`

```{r feature_plt_2, fig.width=15, fig.height=12}

#Pericyte Biomarkers
FeaturePlot(BA4, features = c("PDGFRB",
                              "NOTCH3",
                              "PRKG1",
                              "CSPG4",
                              "GPC5",
                              "RGS5",
                              "PDE7B"
                              ))
```


```{r, fig.width= 7, fig.height=10}
VlnPlot(BA4, features = c("PDGFRB", "NOTCH3", "PRKG1", "CSPG4",  "GPC5", "RGS5", "PDE7B"))
```

* Cluster 0 = Endothelial cells
* Cluster 1 = Pericytes


```{r, include=FALSE}
annot <- "Annotate cells in BA4 endo-pericyte populations"
```
## `r annot`


```{r}
BA4$annotate <- BA4$RNA_snn_res.0.05

Idents(BA4) <- "annotate"

BA4 <- RenameIdents(BA4, "0" = "Endothelial", "1" = "Pericyte")

BA4$annotate <- Idents(BA4)
```


Generate t-SNE with cell annotation
```{r, fig.width=7, fig.height=6, fig.fullwidth=TRUE}
DimPlot(BA4, label = T, repel = T, group.by = "annotate")
```


```{r, message= FALSE}
saveRDS(BA4, "/Users/bonocheong/Desktop/Research_Project/Datasets/Dissertation/data/BA4_annotated.RDS")
```


```{r, message= FALSE}
BA4_EC <- subset(BA4, subset=annotate=="Endothelial")

saveRDS(BA4_EC, "/Users/bonocheong/Desktop/Research_Project/Datasets/Dissertation/data/BA4_EC_annotated.RDS")
```























